DevTools/SourceMaps/explainer.md

```diff
@@ -0,0 +1,245 @@
+
+```
+TypeError: Cannot read properties of undefined (reading 'x')
+    at Render#draw (file1.ts:302:7)
+    at Render#drawLayer (file1.ts:144:5)
+    at GameObject#draw (file2.ts:95:5)
+    at Game#render (file3.ts:178:5)
+    at Game#reflowCanvas (file3.ts:155:5)
+    at [anonymous function passed to addEventListener] (file4.ts:39:5)
+```
+
+But doing this required that I dump the mappings and the source files from the source map and manually inspect the source files. In general, this requires that I use a library to parse the `mappings` field from Source Maps (because [as of Source Maps v3](https://sourcemaps.info/spec.html), this field is stored in a stateful way).
+
+## Proposed Solution
+
+We want to adopt the general approach taken by the `pasta-sourcemaps` ("Pretty (and) Accurate Stack Trace Analysis") library. However, instead of adding to the `names` field, we will add two fields.
+
+* Add two additional field to the source map: `scopes` and `scopeNames`
+* The `scopes` field should be a list of ranges and pointers into `scopeNames`
+* The ranges in the field should always point to **Original Source** locations.
+
+### Practical example
+
+Suppose we have the following TypeScript source:
+
+```ts
+// example.ts
+1 function foo(bar: number): never {
+2     throw new Error('Intentional.');
+3 }
+4 foo();
+```
+
+Direct TypeScript compilation would emit something fairly similar:
+
+```js
+// example.js
+1 function foo(bar) {
+2     throw new Error('Intentional.');
+3 }
+4 foo();
+```
+
+But running it through a minifier would probably transform it to an IIFE lambda:
+
+```js
+// example.min.js
+1 (()=>{throw new Error('Intentional.')})();
+```
+
+Examining the call stack of this looks something like:
+
+```
+Error: Intentional.
+    at <anonymous>:1:13
+    at <anonymous>:1:40
+```
+
+We can get to the exact *locations* with source maps today, but assuming that the `mapping`'s 5th field ("the zero-based index into the `names` list associated with the segment) is encoded as it typically is, which is meant for live debugging, the decoded stack would actually look like this:
+
+```
+Error: Intentional
+    at throw (example.js:2:5)
+    at foo (example.js:4:1)
+```
+
+What we want to get to from here would be, at minimum:
+
+```
+Error: Intentional.
+    at foo (example.js:2:5)
+    at Global code (example.js:4:1)
+```
+
+### Improving the contents of the map
+
+Let us re-examine the original code in this example:
+
+```ts
+1 function foo(bar: number): never {
+2     throw new Error('Intentional.');
+3 }
+4 foo();
+```
+
+The scopes list should be as follows:
+
+1. Line 1 Col 1 -> Line 1 Col 34 (captures `function foo(bar: number): never ` including the trailing space): `Global code`
+1. Line 1 Col 35 -> Line 3 Col 1 (captures the entire body of `foo`): `foo`
+1. Line 3 Col 2 -> Line 4 Col 7 (captures the newline + invocation of `foo`): `Global code`

```

Comment:
Contrasting with pasta-sourcemaps:

Pasta uses "nested" scopes. Referring to ^ as "listed" scopes for this thread, equivalent would be:

1. Line 1 Col 1 -> Line 4 Col 7 (captures until end of invocation of `foo`): `<top-level>`
1. Line 1 Col 35 -> Line 3 Col 1 (captures the entire body of `foo`): `foo`

The [spec](https://github.com/bloomberg/pasta-sourcemaps/blob/9d2022adbd87c76af1d10a352963ff1765c52def/spec.md) doesn't seem to dwell on this, but some observations:

- "nested" is smaller
- Can't recursively bisect the "nested" `scopes` to find position
  - VLQ doesn't permit this anyway, must be linearly unpacked

Reply:
I'm not sure I understand what "nested" scopes are here, but regarding the second comment: the approach to bisecting the `scopes`:

 - Parse all of them
 - Sort by start line, then start column
 - Now you can binary search, you should look for the last entry that starts before and ends after the mapping position

Doesn't that meet your requirements?

Reply:
Cross-post from https://github.com/webpack/webpack/issues/14996#issuecomment-996748709:

> ```
> // *nested*                            // *linear*
> | "<top-level>"                        | "<top-level>"
> |                                      |
> |  |  "causeError"         vs            |  "causeError"
> |  |                                     |
> |                                      | "<top-level>"
> ```

"Nested" scopes have one "entry" per scope in the source code.
"Linear" scopes have an entry each time you transition from one scope to another, left to right, top to bottom.

Nested will tend to result in fewer mappings, and a data structure that more semantically corresponds to the original source

Reply:
Ahhh, I understand! Actually, the way I implemented it in TypeScript was the "nested" version, and while I think I was meaning for this to indicate the "logical" view of the scopes list, but you're right -- I did put this into the expected list of these in the spec. I'll get that fixed.